# **Final Merged Preprocess & Feature Enginnering**

-------

In [2]:
# imports
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder, StandardScaler

import joblib

pd.set_option("display.max_columns", 200)
pd.set_option("display.max_rows", 200)

In [4]:
# load in the data and look at its shape
csv_path = "final_merged.csv"
df = pd.read_csv(csv_path)

print("raw shape:", df.shape)
display(df.head())

raw shape: (2520, 94)


,patientunitstayid,gender,age,ethnicity,admissionheight,admissionweight,unittype,bmi,dialysis,wbc,respiratoryrate,sodium,heartrate,meanbp,ph,hematocrit,creatinine,albumin,pao2,pco2,bun,glucose,bilirubin,fio2,aps_missing,acutephysiologyscore,apachescore,apache_missing,aids,hepaticfailure,lymphoma,metastaticcancer,leukemia,immunosuppression,cirrhosis,diabetes,midur,pred_missing,total_diagnoses,primary_count,major_count,other_count,primary_ratio,major_ratio,pasthistoryvalue,hx_cardio,hx_respiratory,hx_neuro,hx_cancer,hx_renal,hx_liver,hx_endocrine,hx_immuno,hx_heme,hx_none,intaketotal_x,outputtotal_x,dialysistotal_x,nettotal_x,io_missing_x,drug_allergy,non_drug_allergy,hospitalid_x,temp_min,temp_max,hr_min,hr_max,resp_max,sao2_min,sbp_min,dbp_min,map_min,vitals_missing,hr_range,temp_range,intaketotal_y,outputtotal_y,dialysistotal_y,nettotal_y,io_missing_y,hospitalid_y,numbedscategory,teachingstatus,region,hospitaldischargestatus,unitdischargestatus,unitvisitnumber,visitnumber,readmit,ICU_death,hospital_death,ICU_readmit,readmit_clean,bad_outcome
0,141764,Female,87.0,Caucasian,157.5,NaN,Med-Surg ICU,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,36.0,48.0,1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1,0.0,0.0,0.0,0.0,0.000000,0.000000,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,200.0,500.0,0.0,100.0,0,NaN,NaN,59,36.0,37.8,90.0,138.0,29.0,90.0,84.0,44.0,58.0,1,48.0,1.8,200.0,500.0,0.0,100.0,0,59,<100,f,Midwest,Alive,Alive,2,NaN,NaN,0,0,1,0,1
1,141765,Female,87.0,Caucasian,157.5,46.5,Med-Surg ICU,18.745276,0.0,10.2,39.0,139.0,88.0,108.0,-1.0,37.8,1.04,-1.0,-1.0,-1.0,28.0,61.0,-1.0,-1.0,0,23.0,47.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0.0,0.000000,0.000000,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1029.0,600.0,0.0,1029.0,0,NaN,NaN,59,36.0,37.8,72.0,116.0,39.0,95.0,84.0,44.0,58.0,1,44.0,1.8,1029.0,600.0,0.0,1029.0,0,59,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0
2,143870,Male,76.0,Caucasian,167.0,77.5,SICU,27.788734,0.0,11.7,60.0,133.0,40.0,47.0,-1.0,34.1,1.14,-1.0,-1.0,-1.0,14.0,140.0,-1.0,-1.0,0,43.0,60.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,3.0,1.0,1.0,1.0,0.333333,0.333333,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1556.0,300.0,0.0,1556.0,0,NaN,NaN,68,36.0,37.8,40.0,55.0,86.0,80.0,53.0,36.0,47.0,1,15.0,1.8,1556.0,300.0,0.0,1556.0,0,68,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0
3,144815,Female,34.0,Caucasian,172.7,60.3,Med-Surg ICU,20.217741,0.0,7.9,6.0,141.0,131.0,61.0,-1.0,36.6,0.63,3.6,-1.0,-1.0,6.0,82.0,0.5,-1.0,0,25.0,25.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0.0,0.000000,0.000000,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1330.0,900.0,0.0,1329.0,0,NaN,NaN,56,36.0,37.8,68.0,118.0,42.0,96.0,84.0,44.0,58.0,1,50.0,1.8,1330.0,900.0,0.0,1329.0,0,56,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0
4,145427,Male,61.0,Caucasian,177.8,91.7,SICU,29.007201,0.0,21.1,41.0,141.0,49.0,72.0,-1.0,40.4,1.05,-1.0,-1.0,-1.0,14.0,118.0,-1.0,-1.0,0,26.0,37.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,8.0,3.0,3.0,2.0,0.375000,0.375000,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1000.0,1000.0,0.0,1000.0,0,NaN,NaN,68,36.0,37.8,49.0,85.0,41.0,89.0,84.0,44.0,58.0,1,36.0,1.8,1000.0,1000.0,0.0,1000.0,0,68,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0


In [6]:
# to check cols & target

# print all col names 
print(df.columns.tolist())

# target counts
print("\ntarget distribution:")
print(df["bad_outcome"].value_counts(dropna=False))

# look at target proportions
print("\ntarget proportions:")
print(df["bad_outcome"].value_counts(normalize=True, dropna=False))

['patientunitstayid', 'gender', 'age', 'ethnicity', 'admissionheight', 'admissionweight', 'unittype', 'bmi', 'dialysis', 'wbc', 'respiratoryrate', 'sodium', 'heartrate', 'meanbp', 'ph', 'hematocrit', 'creatinine', 'albumin', 'pao2', 'pco2', 'bun', 'glucose', 'bilirubin', 'fio2', 'aps_missing', 'acutephysiologyscore', 'apachescore', 'apache_missing', 'aids', 'hepaticfailure', 'lymphoma', 'metastaticcancer', 'leukemia', 'immunosuppression', 'cirrhosis', 'diabetes', 'midur', 'pred_missing', 'total_diagnoses', 'primary_count', 'major_count', 'other_count', 'primary_ratio', 'major_ratio', 'pasthistoryvalue', 'hx_cardio', 'hx_respiratory', 'hx_neuro', 'hx_cancer', 'hx_renal', 'hx_liver', 'hx_endocrine', 'hx_immuno', 'hx_heme', 'hx_none', 'intaketotal_x', 'outputtotal_x', 'dialysistotal_x', 'nettotal_x', 'io_missing_x', 'drug_allergy', 'non_drug_allergy', 'hospitalid_x', 'temp_min', 'temp_max', 'hr_min', 'hr_max', 'resp_max', 'sao2_min', 'sbp_min', 'dbp_min', 'map_min', 'vitals_missing', 'hr_

In [10]:
def audit_dataframe(dataframe):
    # summary table, ea row is a col
    summary = pd.DataFrame({
        "dtype": dataframe.dtypes.astype(str),
        # percentage of missing 
        "missing_pct": dataframe.isna().mean().round(3),
        #unique values?
        "n_unique_including_nan": dataframe.nunique(dropna=False),
    })

    # find the most common value and its frequency (as a %)
    summary["top_value_pct"] = [
        dataframe[col]                          
        .value_counts(dropna=False, normalize=True)  
        .iloc[0] # most frequent 
        .round(3)                               
        for col in dataframe.columns
    ]
    return summary.sort_values(["dtype", "n_unique_including_nan", "missing_pct"])


audit = audit_dataframe(df)
display(audit)


# checks

print("\nobject columns:")

display(
    df.select_dtypes(include="object")
      .nunique(dropna=False)
      .sort_values()
)


print("\nnumeric columns containing -1:")

# dictionary to track how many times -1 appears in the num cols
neg1_counts = {
    col: int((df[col] == -1).sum())  
    for col in df.select_dtypes(include="number").columns  
    if (df[col] == -1).any()  
}

display(pd.Series(neg1_counts).sort_values(ascending=False))

,dtype,missing_pct,n_unique_including_nan,top_value_pct
aids,float64,0.000,1,1.000
hepaticfailure,float64,0.000,2,0.989
lymphoma,float64,0.000,2,0.994
metastaticcancer,float64,0.000,2,0.981
leukemia,float64,0.000,2,0.992
immunosuppression,float64,0.000,2,0.977
cirrhosis,float64,0.000,2,0.988
diabetes,float64,0.000,2,0.806
midur,float64,0.000,2,0.994
hx_cardio,float64,0.000,2,0.692



object columns:


teachingstatus              2
gender                      3
hospitaldischargestatus     3
unitdischargestatus         3
numbedscategory             5
region                      5
ethnicity                   7
unittype                    8
pasthistoryvalue           67
dtype: int64


numeric columns containing -1:


pco2                    1800
fio2                    1800
pao2                    1800
ph                      1800
bilirubin               1442
albumin                 1351
wbc                      688
hematocrit               638
bun                      565
creatinine               561
sodium                   555
glucose                  388
meanbp                    60
respiratoryrate           57
heartrate                 42
acutephysiologyscore      41
apachescore               41
nettotal_x                 8
nettotal_y                 8
map_min                    1
dtype: int64

-----------
### **fix duplicate _x/_y columns first**
- in the merged file, there's several pairs are exact duplicates?

In [13]:
df_clean = df.copy()

duplicate_pairs = [
    ("intaketotal_x", "intaketotal_y", "intaketotal"),
    ("outputtotal_x", "outputtotal_y", "outputtotal"),
    ("dialysistotal_x", "dialysistotal_y", "dialysistotal"),
    ("nettotal_x", "nettotal_y", "nettotal"),
    ("io_missing_x", "io_missing_y", "io_missing"),
    ("hospitalid_x", "hospitalid_y", "hospitalid"),
]

for left_col, right_col, new_name in duplicate_pairs:
    if left_col in df_clean.columns and right_col in df_clean.columns:
        same_values = (df_clean[left_col].fillna("__NA__") == df_clean[right_col].fillna("__NA__")).all()
        print(f"{left_col} vs {right_col} identical? -> {same_values}")
        if same_values:
            df_clean = df_clean.rename(columns={left_col: new_name})
            df_clean = df_clean.drop(columns=[right_col])
        else:
            print(f"warning!!: {left_col} and {right_col}  not identical. review again pls")

print("\nthe shape after duplicate cleanup:", df_clean.shape)
display(df_clean.head())

intaketotal_x vs intaketotal_y identical? -> True
outputtotal_x vs outputtotal_y identical? -> True
dialysistotal_x vs dialysistotal_y identical? -> True
nettotal_x vs nettotal_y identical? -> True
io_missing_x vs io_missing_y identical? -> True
hospitalid_x vs hospitalid_y identical? -> True

the shape after duplicate cleanup: (2520, 88)


,patientunitstayid,gender,age,ethnicity,admissionheight,admissionweight,unittype,bmi,dialysis,wbc,respiratoryrate,sodium,heartrate,meanbp,ph,hematocrit,creatinine,albumin,pao2,pco2,bun,glucose,bilirubin,fio2,aps_missing,acutephysiologyscore,apachescore,apache_missing,aids,hepaticfailure,lymphoma,metastaticcancer,leukemia,immunosuppression,cirrhosis,diabetes,midur,pred_missing,total_diagnoses,primary_count,major_count,other_count,primary_ratio,major_ratio,pasthistoryvalue,hx_cardio,hx_respiratory,hx_neuro,hx_cancer,hx_renal,hx_liver,hx_endocrine,hx_immuno,hx_heme,hx_none,intaketotal,outputtotal,dialysistotal,nettotal,io_missing,drug_allergy,non_drug_allergy,hospitalid,temp_min,temp_max,hr_min,hr_max,resp_max,sao2_min,sbp_min,dbp_min,map_min,vitals_missing,hr_range,temp_range,numbedscategory,teachingstatus,region,hospitaldischargestatus,unitdischargestatus,unitvisitnumber,visitnumber,readmit,ICU_death,hospital_death,ICU_readmit,readmit_clean,bad_outcome
0,141764,Female,87.0,Caucasian,157.5,NaN,Med-Surg ICU,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1,36.0,48.0,1,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1,0.0,0.0,0.0,0.0,0.000000,0.000000,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,200.0,500.0,0.0,100.0,0,NaN,NaN,59,36.0,37.8,90.0,138.0,29.0,90.0,84.0,44.0,58.0,1,48.0,1.8,<100,f,Midwest,Alive,Alive,2,NaN,NaN,0,0,1,0,1
1,141765,Female,87.0,Caucasian,157.5,46.5,Med-Surg ICU,18.745276,0.0,10.2,39.0,139.0,88.0,108.0,-1.0,37.8,1.04,-1.0,-1.0,-1.0,28.0,61.0,-1.0,-1.0,0,23.0,47.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0.0,0.000000,0.000000,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1029.0,600.0,0.0,1029.0,0,NaN,NaN,59,36.0,37.8,72.0,116.0,39.0,95.0,84.0,44.0,58.0,1,44.0,1.8,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0
2,143870,Male,76.0,Caucasian,167.0,77.5,SICU,27.788734,0.0,11.7,60.0,133.0,40.0,47.0,-1.0,34.1,1.14,-1.0,-1.0,-1.0,14.0,140.0,-1.0,-1.0,0,43.0,60.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,3.0,1.0,1.0,1.0,0.333333,0.333333,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1556.0,300.0,0.0,1556.0,0,NaN,NaN,68,36.0,37.8,40.0,55.0,86.0,80.0,53.0,36.0,47.0,1,15.0,1.8,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0
3,144815,Female,34.0,Caucasian,172.7,60.3,Med-Surg ICU,20.217741,0.0,7.9,6.0,141.0,131.0,61.0,-1.0,36.6,0.63,3.6,-1.0,-1.0,6.0,82.0,0.5,-1.0,0,25.0,25.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,0.0,0.0,0.0,0.0,0.000000,0.000000,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1330.0,900.0,0.0,1329.0,0,NaN,NaN,56,36.0,37.8,68.0,118.0,42.0,96.0,84.0,44.0,58.0,1,50.0,1.8,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0
4,145427,Male,61.0,Caucasian,177.8,91.7,SICU,29.007201,0.0,21.1,41.0,141.0,49.0,72.0,-1.0,40.4,1.05,-1.0,-1.0,-1.0,14.0,118.0,-1.0,-1.0,0,26.0,37.0,0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0,8.0,3.0,3.0,2.0,0.375000,0.375000,no health problems,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0,1000.0,1000.0,0.0,1000.0,0,NaN,NaN,68,36.0,37.8,49.0,85.0,41.0,89.0,84.0,44.0,58.0,1,36.0,1.8,<100,f,Midwest,Alive,Alive,1,1.0,0.0,0,0,0,0,0


### **convert sentinel '-1' values to missing values**
- had to do some research but a lot of the clinical variables use '-1' as a placeholder for “not measured / not available ” info.  
- so not a real physiologic value, so we convert those to nan

In [16]:

numeric_cols = df_clean.select_dtypes(include="number").columns.tolist()

neg1_before = {
    col: int((df_clean[col] == -1).sum())
    for col in numeric_cols
    if (df_clean[col] == -1).any()
}

for col in numeric_cols:
    if (df_clean[col] == -1).any():
        df_clean[col] = df_clean[col].replace(-1, np.nan)

neg1_after = {
    col: int((df_clean[col] == -1).sum())
    for col in df_clean.select_dtypes(include="number").columns
    if (df_clean[col] == -1).any()
}

print("columns that had -1:")
display(pd.Series(neg1_before).sort_values(ascending=False))

print("\nremaining cols still containing -1 (should be empty or very small if truly valid):")
display(pd.Series(neg1_after))

columns that had -1:


pao2                    1800
fio2                    1800
pco2                    1800
ph                      1800
bilirubin               1442
albumin                 1351
wbc                      688
hematocrit               638
bun                      565
creatinine               561
sodium                   555
glucose                  388
meanbp                    60
respiratoryrate           57
heartrate                 42
acutephysiologyscore      41
apachescore               41
nettotal                   8
map_min                    1
dtype: int64


remaining cols still containing -1 (should be empty or very small if truly valid):


Series([], dtype: object)

----------------------------------------------
### **target and drop leakage/bloat cols**

why these drops?
- **ids** are not useful predictors
- **discharge/death/ readmit outcome fields** directly reveal the target or happen after the prediction moment
- raw **`pasthistoryvalue`** is the main source of dummy explosion
- we already keep the cleaner grouped history flags like `hx_cardio`, `hx_renal`, etc.
- visit counters can leak future trajectory for readmission-style outcomes

In [19]:
target_col = "bad_outcome"

drop_columns = [
    # ids
    "patientunitstayid",

    # raw history text/high-cardinality 
    "pasthistoryvalue",

    # post-outcome/leakage variables
    "hospitaldischargestatus",
    "unitdischargestatus",
    "readmit",
    "ICU_death",
    "hospital_death",
    "ICU_readmit",
    "readmit_clean",

    # likely downstream or trajectory-style counters
    "visitnumber",
    "unitvisitnumber",
]

existing_drop_columns = [col for col in drop_columns if col in df_clean.columns]

model_df = df_clean.drop(columns=existing_drop_columns)

print("dropped columns:")
print(existing_drop_columns)

print("\nshape after manual drops:", model_df.shape)

dropped columns:
['patientunitstayid', 'pasthistoryvalue', 'hospitaldischargestatus', 'unitdischargestatus', 'readmit', 'ICU_death', 'hospital_death', 'ICU_readmit', 'readmit_clean', 'visitnumber', 'unitvisitnumber']

shape after manual drops: (2520, 77)


In [21]:
## another missinginess audit

missing_counts = model_df.isnull().sum()
missing_percent = (missing_counts / len(model_df)) * 100

missing_summary = (
    pd.DataFrame({
        "missing_count": missing_counts,
        "missing_percent": missing_percent
    })
    .sort_values(by="missing_percent", ascending=False)
)

# show only columns with missing values
missing_summary = missing_summary[missing_summary["missing_count"] > 0]

print("missingness sorted:")
display(missing_summary)

missingness sorted:


,missing_count,missing_percent
ph,2115,83.928571
fio2,2115,83.928571
pco2,2115,83.928571
pao2,2115,83.928571
bilirubin,1757,69.722222
non_drug_allergy,1746,69.285714
drug_allergy,1746,69.285714
albumin,1666,66.111111
wbc,1003,39.801587
hematocrit,953,37.817460


In [23]:
# colsw/ >30% missing
high_missing = missing_summary[missing_summary["missing_percent"] > 30]

print("columns with >30% missing:")
display(high_missing)

columns with >30% missing:


,missing_count,missing_percent
ph,2115,83.928571
fio2,2115,83.928571
pco2,2115,83.928571
pao2,2115,83.928571
bilirubin,1757,69.722222
non_drug_allergy,1746,69.285714
drug_allergy,1746,69.285714
albumin,1666,66.111111
wbc,1003,39.801587
hematocrit,953,37.817460


####  **missingness obs & things to consider later**

- **ph, fio2, pco2, pao2** all have **~83.9% missing**

#### thought process
hm instead of treating this as “bad data,” we can.. ABG (Arterial Blood Gas)
- if ABG is measured, patient is likely more critical
- if not measured, patient may be more stable

### ideas to apply later (feature engineering)

#### abg signal (important, binary)
- make a flag:
  - is_abg_measured = 1 if any of ph, fio2, pco2, pao2 present
  - else 0
- could act as a **severity indicator**

- a consideration:
  - raw ABG values are very sparse, might not be useful directly
  - the *presence* of the test may be more predictive than the values
----------

#### **remove constant cols**
constant columns add nothing to modeling and only clutter the feature space

In [30]:
X_full = model_df.drop(columns=[target_col]).copy()
y_full = model_df[target_col].copy()

constant_cols = [col for col in X_full.columns if X_full[col].nunique(dropna=False) <= 1]

print("constant columns found:", len(constant_cols))
print(constant_cols)

X_full = X_full.drop(columns=constant_cols)

print("\nfeature matrix shape after constant drop:", X_full.shape)
print("target distribution:")
display(y_full.value_counts(dropna=False))
display(y_full.value_counts(normalize=True).round(3))

constant columns found: 1
['aids']

feature matrix shape after constant drop: (2520, 75)
target distribution:


bad_outcome
0    1938
1     582
Name: count, dtype: int64

bad_outcome
0    0.769
1    0.231
Name: proportion, dtype: float64

In [32]:
### that will do for now... save into csv
clean_path = "simple_preprocess.csv"
model_df.to_csv(clean_path, index=False)
print(f"saved data to: {clean_path}")

saved data to: simple_preprocess.csv
